Subbus/Classes/Subbus.PersistentEvent.swift

```diff
@@ -0,0 +1,140 @@
+        /// will receive the event, then it will be cleared.
+        case clearAfterAllCurrentSubscribersNotified
+        /// Event will *never* be cleared, regardless of what subscribers return.
+        case neverClear
+    }
+    
+    /// The event's persistence rule.
+    var persistenceRule: PersistenceRule
+    
+    /// Initializer that requires a persistence rule.
+    ///
+    /// Parameters:
+    ///     - persistenceRule: the rule of persistence.
+    init(persistenceRule: PersistenceRule) {
+        self.persistenceRule = persistenceRule
+    }
+}
+
+/// The protocol defining the persistence capabilities added to Subbus.
+protocol SubbusPersistentEventProtocol {
+    /// Adds a new subscription to Subbus with a `PersistentEvent`.
+    ///
+    /// Parameters:
+    ///     - id: The id of the subscriber to this event.  Used for unregistering.
+    ///     - event: The event type to register the subscriber to.  This class will be a subclass of `PersistentEvent`.
+    ///     - replace: Whether this event should replace a previous subscription with the same `id` and `event`.
+    ///     - callback: The handler block for when the event is `Post`ed.
+    static func addSubscription<I, T: PersistentEvent>(id: I, event: T.Type, replace: Bool, callback: @escaping (T) -> PersistentEvent.HandlerResult)
+    
+    /// Posts a `PersistentEvent` to the subscribed receivers.
+    ///
+    /// Parameters:
+    ///     - event: The event to be posted.
+    static func post<T: PersistentEvent>(event: T)
+}
+
+/// A convenience extension to add shortened functions to the `SubbusPersistentEventProtocol`.
+extension SubbusPersistentEventProtocol {
+    /// Adds a new subscription to Subbus with a `PersistentEvent`.
+    ///
+    /// Parameters:
+    ///     - id: The id of the subscriber to this event.  Used for unregistering.
+    ///     - event: The event type to register the subscriber to.
+    ///     - callback: The handler block for when the event is `Post`ed.
+    public static func addSubscription<I, T: PersistentEvent>(id: I, event: T.Type, callback: @escaping (T) -> PersistentEvent.HandlerResult) {
+        addSubscription(id: id, event: event, replace: false, callback: callback)
+    }
+}
+
+/// Conformance to the `SubbusPersistentEventProtocol`.
+extension Subbus: SubbusPersistentEventProtocol {
+    public static func addSubscription<I, T: PersistentEvent>(id: I, event: T.Type, replace: Bool, callback: @escaping (T) -> PersistentEvent.HandlerResult) {
+        let eventType = String(reflecting: T.self)
+        
+        //Register listener
+        guard let subscription = Subscription(identifier: id, eventType: eventType, handler: callback) else { return }
+        addSubscription(subscription, replace: replace)
+        
+        // Post existing events to this new subscription
+        for index in stride(from: unhandledEvents.count - 1, through: 0, by: -1)  {
```

Comment:
I'm very confused by what is happening here (the entire loop, not just the stride logic)